### DISCUSSION TITLE
Can not find `fs`

### DISCUSSION THREAD

**Discussion by ghost** [webpack/webpack#16656] [Q&A]
**What is the current behavior?**

I am using this line in my script:

```js
import {readFile} from "fs/promises";
```

And it errors out with this log:

<summary> 
  Log 
</summary>

<details>

```shell

ERROR in ./src/ts/index.ts 41:13-26
Module not found: Error: Can't resolve 'fs' in '/bundle/src/ts'
resolve 'fs' in '/bundle/src/ts'
  Parsed request is a module
  using description file: /bundle/package.json (relative path: ./src/ts)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /bundle/src/ts/node_modules doesn't exist or is not a directory
      /bundle/src/node_modules doesn't exist or is not a directory
      looking for modules in /bundle/node_modules
        single file module
          using description file: /bundle/package.json (relative path: ./node_modules/fs)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /bundle/node_modules/fs doesn't exist

```

</details>

**If the current behavior is a bug, please provide the steps to reproduce.**

Remains to add my config:

```js
const path = require("path");
//const webpack = require('webpack'); //to access built-in plugins

const webConfig = {
  target: "web",
  mode: "development",
  entry: { main: ["./src/ts/index.ts"] },
  output: {
    filename: "[name].web.js",
    path: path.resolve(__dirname, "dist"),
  },
  optimization: {
    splitChunks: {
      chunks: "all",
    },
  },
  watch: true,
  watchOptions: {
    ignored: /node_modules/,
  },
  module: {
    rules: [
      {
        test: /\.ts$/i,
        use: "ts-loader",
      },
    ],
  },
  // this uses the index.html, injects the scripts and places it on /dist
  //plugins: [new HtmlWebpackPlugin({ template: "index.html" })],
  resolve: {
    extensions: [".ts", ".js"],
//    modules: ["node_modules"],
  },
};

const nodeConfig = {
  ...webConfig,
  target: "node",
  output: { filename: "[name].node.js", path: path.resolve(__dirname, "dist") },
};

module.exports = [webConfig , nodeConfig];

```

**What is the expected behavior?**

Load readFile, or skip it.

**Other relevant information:**
webpack version:  5
Node.js version: 19
Operating System: Ubuntu
Additional tools: ts-loader

**Comment by ghost** [comment]
I've noticed a similar issue here https://github.com/webpack-contrib/css-loader/issues/447 unsolved but at least adding this in the webpack config helped to compile:

```js
  resolve: {
    fallback: {
      fs: false,
    },
  },
```

it ignores the fs for browsers, and searches it for node. I wonder if you have any other solution to recommend ?

**Reply by alexander-akait** [reply]
You have `target: "web"`, `fs` module doesn't exits in browsers, that is why you got the message

**Reply by ghost** [reply]
> You have `target: "web"`, `fs` module doesn't exits in browsers, that is why you got the message

But that is in purpose, I need a single script that does something like this:

```ts
if(isNode) loadFs()
else if (isBrowser) loadOtherStuff()
```

And it was erroring out, not only warning. 

I have seen this pattern in other programs but can you tell me if this isn't correct in your opinion ?

Thanks for your reply.

**Reply by alexander-akait** [reply]
You want universal app, I see, I think better to use multi compiler mode with different targets, you will have two files, but it will have less size for the browser, because terser/minifier removed dead code for node.js, feel free to feedback

**Reply by ghost** [reply]
Thank you. So I'd rather have a `browser` key in the `package.json` and then create 2 separate files one for node, one for browser correct ? @alexander-akait

**Reply by alexander-akait** [reply]
Yes, I would recommend it, less size and better to catch universal app/library problems

**Reply by ghost** [reply]
Thanks for your help
